### ISSUE TITLE
torch needs to SHOW that it support sm_89 even if functionally the same as sm_86

### ISSUE THREAD

**Issue by aifartist** [pytorch/pytorch#95648]
### 🐛 Describe the bug

torch needs to SHOW that it supports sm_89 even if functionally the same as sm_86.
While it may be true that optimizing specifically for an sm_89 provides no benefits beyond what optimizing for an sm_86 does, by NOT listing sm_89 in either what torch._C._cuda_getArchFlags() nor torch.cuda.get_arch_list() outputs, code which checks if the actual GPU(in my case (8,9)) is in the list of supported GPU's can lead to "generic" optimization which are far from optimal.

If the left hand of pytorch say you have a sm_89 and the right hand say it supports "sm_37 sm_50 sm_60 sm_70 sm_75 sm_80 sm_86 sm_90" with sm_89 being missing from the list this is a problem and I have seen code where this kind of lookup was done.  If you want to save code space and, as part of the build, not generate sm_89 code which is identical to the sm_86 then use "if" tests to redirect sm_89 to the sm_86 code paths to keep end user code from needing to be aware of this quirk. But you need to list sm_89 as a supported architecture.

NOTE: If someone does a single TORCH_CUDA_ARCH_LIST build of sm_86 OR sm_89 both should return "sm_86 sm_89" and ['sm_86', 'sm_89'] from the two functions shown above.  However, I won't push on that.  Anyone doing their own personal build of pytorch probably doesn't need this.  It is the end user or basic app coders that are using the official FULL build of pytorch that needs ALL supported GPU's to be listed because "support for a GPU" is not the same as "does a GPU have any unique performance features another GPU earlier in the list has".

```
torch._C._cuda_getArchFlags() = sm_37 sm_50 sm_60 sm_70 sm_75 sm_80 sm_86 sm_90
torch.cuda.get_arch_list() = ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86', 'sm_90']
```

### Versions

Not needed

**Comment by ptrblck** [comment]
The return values from `torch.cuda.get_arch_list()` indicate return the GPU architectures the binary was **compiled** for as given in the [docs](https://pytorch.org/docs/stable/generated/torch.cuda.get_arch_list.html#torch-cuda-get-arch-list) and the missing `sm_89` is correct.
Using this call to check for **supported** architectures is wrong. 
However, I see the validity of this use case and we might want to discuss a new API returning supported architectures as an addition. 

CC @ngimel

**Comment by aifartist** [comment]
Ok.  I didn't realize "the letter of the law" regarding get_arch_list().
Forgetting about user code that mistakenly understood what this did I ask:
It it guaranteed that all torch internal code when asking for <something ... sm_89 ... something> will give you sm_86 capabilities and NOT revert to some lowest common denominator of capability because it wasn't built with sm_89?

I discovered in the last week that GCC 11.3 when using "-mtune=native" it was doing "-mtune=generic" because it didn't recognize RaptorLake.  GCC 13 will, when it comes out, support RaptorLake.  FYI, with "-march=native" it at least translates that to "-march=alderlake" so some benefits can be had in terms of "instruction set".  But generic **tuning** is for ancient CPU's.  My work around is to use "-mtune=alderlake".  Some things in the build of pytorch didn't like trying to use a newer GCC than 11x.

**Comment by ngimel** [comment]
Yes, it is guaranteed. CUDA first tries to find closest binary-compatible code for sm_89 (in this case it'll be sm_86), if binary-compatible code is not found it will jit compile the code from ptx if ptx exists, if ptx doesn't exist it will error. Pytorch, when compiled with `sm_86` compilation flag, packages sm_86 kernels and they will be used on sm_89. Torch internal code is not asking for specific compute capabilities, these things are handled by cuda driver.
